File: src/main/java/aoc2018/Day21.kt
package aoc2018


fun main() = Day21().run()

class Day21 : Day(21) {
    override fun part1(input: List<String>): Any {
        val ipCom = input[0].substringAfter(' ').toInt()
        val instructions = input.subList(1, input.size).map<String, (IntArray) -> Unit> { line ->
            val (a, b, c) = line.substringAfter(' ').split(' ').map(String::toInt)
            val opcode = line.substringBefore(' ')
            val command = commands[opcode]!!
            return@map { command(it, a, b, c) }
        }
        val registers = IntArray(6)
        var ip = 0

        while ((0 <= ip) && (ip < instructions.size)) {
            instructions[ip](registers)
            ip = ++registers[ipCom]
            if (ip == 28) {
                break
            }
        }
        return registers[3]
    }

    override fun part2(input: List<String>): Any {
        val ipCom = input[0].substringAfter(' ').toInt()
        val instructions = input.subList(1, input.size).map<String, (IntArray) -> Unit> { line ->
            val (a, b, c) = line.substringAfter(' ').split(' ').map(String::toInt)
            val opcode = line.substringBefore(' ')
            val command = commands[opcode]!!
            return@map { command(it, a, b, c) }
        }
        val registers = IntArray(6)
        var ip = 0
        val seen = mutableSetOf<Int>()
        var last = 0

        while ((0 <= ip) && (ip < instructions.size)) {
            instructions[ip](registers)
            ip = ++registers[ipCom]
            if (ip == 28) {
                if (!seen.add(registers[3]))
                    break
                last = registers[3]
            }
        }
        return last
    }
}

File: src/main/java/aoc2022/Day21.scala
package aoc2022

import java.math.MathContext
import scala.collection.mutable

object Day21 extends Day(21) {

  private val monkeyOps: Map[String, () => BigDecimal] = input.map(line => (line.take(4), expression(line.drop(6)))).toMap
  private val monkeyAnswers: mutable.Map[String, BigDecimal]
  = mutable.HashMap.empty[String, BigDecimal]

  def main(args: Array[String]): Unit = run()

  override def part1(): Any = {
    findAnswers()
    monkeyAnswers("root").toBigInt
  }

  override def part2(): Any = {
    val equalMonkeys = input.find(_.startsWith("root")).get.drop(6).split(" [+\\-*/] ")
    val left = equalMonkeys(0)
    val right = equalMonkeys(1)
    findAnswers(0)
    val left0 = monkeyAnswers(left)
    val right0 = monkeyAnswers(right)
    findAnswers(1)
    val left1 = monkeyAnswers(left)
    val right1 = monkeyAnswers(right)
    if (left0 == left1) {
      val slope = right1 - right0
      ((left0 - right0) / slope).toBigInt
    } else if(right1 == right0){
      val slope = left1 - left0
      ((right0 - left0) / slope).toBigInt
    }else{
      throw new IllegalStateException()
    }
  }

  private def findAnswers(humn: BigDecimal = monkeyOps("humn").apply()): Unit = {
    monkeyAnswers.clear()
    monkeyAnswers("humn") = humn
    while (monkeyAnswers.size < monkeyOps.size) {
      monkeyOps.view.filterKeys(!monkeyAnswers.contains(_)).foreach((name, exp) => {
        try {
          monkeyAnswers(name) = exp.apply()
        } catch {
          case _: NoSuchElementException =>
          case e: Throwable => throw e
        }
      })
    }
  }

  private def expression(exp: String): () => BigDecimal = {
    if (exp.head.isDigit) {
      () => BigDecimal(exp)
    } else {
      val split = exp.split(' ')
      () => {
        val a = monkeyAnswers(split(0))
        val b = monkeyAnswers(split(2))
        split(1).head match {
          case '+' => a + b
          case '-' => a - b
          case '*' => a * b
          case '/' => a / b
        }
      }
    }
  }
}
